sdk/keyvault/Azure.Security.KeyVault.Certificates/src/LightweightPkcs8Decoder.cs

```diff
@@ -0,0 +1,192 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Security.Cryptography;
+
+namespace Azure.Core
+{
+    internal partial class LightweightPkcs8Decoder
+    {
+        private static readonly byte[] s_derIntegerOne = { 0x02, 0x01, 0x01 };
+
+        private static readonly byte[] s_ecAlgorithmId =
+        {
+            0x06, 0x07, 0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x02, 0x01, // OID 1.2.840.10045.2.1, id-ecPublicKey
+        };
+
+        public static ECDsa DecodeECDsaPkcs8(byte[] pkcs8Bytes)
+        {
+            // Based on https://tools.ietf.org/html/rfc5208 and https://www.secg.org/sec1-v2.pdf
+
+            int offset = 0;
+
+            // PrivateKeyInfo SEQUENCE
+            ConsumeFullPayloadTag(pkcs8Bytes, ref offset, 0x30);
+
+            // PKCS#8 PrivateKeyInfo.version == 0
+            ConsumeMatch(pkcs8Bytes, ref offset, s_derIntegerZero);
+
+            // PKCS#8 PrivateKeyInfo.sequence { AlgorithmIdentifier, CurveName }
+            ReadPayloadTagLength(pkcs8Bytes, ref offset, 0x30);
+
+            // id-ecPublicKey AlgorithmIdentifier value
+            ConsumeMatch(pkcs8Bytes, ref offset, s_ecAlgorithmId);
+
+            // CurveName: may be OID or within a sequence from Key Vault.
+            string curveNameOid;
+            if (pkcs8Bytes[offset] == 0x06)
+            {
+                // Get OID for P-256K, P384, and P-521 from Key Vault.
+                curveNameOid = ReadObjectIdentifier(pkcs8Bytes, ref offset);
+            }
+            else if (pkcs8Bytes[offset] == 0x30)
+            {
+                // Get OID for P-256 from Key Vault.
+                int sequenceLength = ReadPayloadTagLength(pkcs8Bytes, ref offset, 0x30);
+                int innerOffset = offset;
+
+                // version == 1
+                ConsumeMatch(pkcs8Bytes, ref innerOffset, s_derIntegerOne);
+
+                // SEQUENCE
+                ReadPayloadTagLength(pkcs8Bytes, ref innerOffset, 0x30);
+
+                // curveName
+                curveNameOid = ReadObjectIdentifier(pkcs8Bytes, ref innerOffset);
+
+                // The prime-field OID is used by Key Vault for P-256K but is not supported by .NET.
+                // Instead, use the secp256k1 OID which is also used by Azure.Security.KeyVault.Keys.
+                if ("1.2.840.10045.1.1".Equals(curveNameOid, StringComparison.Ordinal))
+                {
+                    curveNameOid = "1.3.132.0.10";
+                }
```

Comment:
This is, as discussed in chat, a bogus assertion.  It may hold true in your scoped usage, but this really means "read the manual ECCurve parameters from the rest of the structure."  It's not the named curve 1.3.123.0.10, but rather a "here are some curve parameters, trust me that they're trustworthy (wink, wink)"

Reply:
Agreed. I'm working on this change now. I wanted to get the PR initially reviewed since not much code will change for this change we discussed (rather, the code will be rather limited to this file and related tests).